### PR TITLE
Switch to factory methods

### PR DESCRIPTION
**Pull request by siy**

### CODE REVIEW

File: src/main/java/com/radixdlt/crypto/ECPublicKey.java [hunk 1]
```diff
@@ -41,52 +41,24 @@
...
-		case 4:
-			if (publicKey.length != (BYTES * 2) + 1) {
-				throw new CryptoException("Public key is an invalid uncompressed size");
-			}
-			break;
-		default:
-			throw new CryptoException("Public key is an invalid format");
-		}
+	@JsonCreator
+	public static ECPublicKey fromBytes(byte[] key) throws CryptoException {
```
Comment (Sajjon) [inline src/main/java/com/radixdlt/crypto/ECPublicKey.java]: I thought we were going to get rid of `CryptoException`? I would like that, and replace with specific purpose `PublicKeyCreationException` (in this case)

File: src/main/java/com/radixdlt/crypto/ECKeyUtils.java [hunk 2]
```diff
@@ -125,6 +125,29 @@ static void validatePrivate(byte[] privateKey) throws CryptoException {
 		}
 	}
 
+	static void validatePublic(byte[] publicKey) throws CryptoException {
```
Comment (Sajjon) [inline src/main/java/com/radixdlt/crypto/ECKeyUtils.java]: Replace `CryptoException` with a new `PublicKeyValidationException` ?

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java [hunk 3]
```diff
@@ -109,15 +93,26 @@ public static ECKeyPair fromSeed(byte[] seed) {
...
 		}
 	}
 
+	/**
+	 * Restore {@link ECKeyPair} instance from given private key by computing corresponding public key.
+	 * @param privateKey byte array which contains private key.
+	 * @return A keypair for provided private key
+	 * @throws CryptoException
+	 */
+	public static ECKeyPair fromPrivateKey(byte[] privateKey) throws CryptoException {
```
Comment (Sajjon) [inline src/main/java/com/radixdlt/crypto/ECKeyPair.java]: Replace `CryptoException` with a new `KeyPairCreationExceptions` or similar?

**Review by Sajjon — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 4]
```diff
@@ -102,6 +102,17 @@ jacocoTestReport {
     }
 }
 
+configurations.all {
```
Comment (clstrfsck) [inline build.gradle:105]: This is a bit ugly, as we need to do it in all upstream repos, but I can't really think of a better way.

File: src/test/java/com/radixdlt/crypto/ECKeyUtilsTest.java [hunk 5]
```diff
@@ -83,4 +83,63 @@ public void testAdjustArrayFail() {
 		Arrays.fill(testArray, (byte) 1);
 		ECKeyUtils.adjustArray(testArray, ECKeyPair.BYTES);
 	}
+
+	@Test(expected = CryptoException.class)
```
Comment (clstrfsck) [inline src/test/java/com/radixdlt/crypto/ECKeyUtilsTest.java]: Extra tests here are good.

**Review by clstrfsck — APPROVED** [review]
Looks good to me.

File: src/main/java/com/radixdlt/crypto/ECPublicKey.java (hunk 1 above)
Comment (siy) [reply src/main/java/com/radixdlt/crypto/ECPublicKey.java]: Done

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java (hunk 3 above)
Comment (siy) [reply src/main/java/com/radixdlt/crypto/ECKeyPair.java]: Done

File: src/test/java/com/radixdlt/crypto/ECKeyUtilsTest.java (hunk 5 above)
Comment (siy) [reply src/test/java/com/radixdlt/crypto/ECKeyUtilsTest.java]: Thanks!

File: src/main/java/com/radixdlt/crypto/ECKeyUtils.java (hunk 2 above)
Comment (siy) [reply src/main/java/com/radixdlt/crypto/ECKeyUtils.java]: Done

File: build.gradle (hunk 4 above)
Comment (siy) [reply build.gradle:105]: Yeah, I've tried a number of different options, but all of them look almost as bad. One possible alternative is to repackage artifacts, but this has significant overhead and therefore has sense to bigger teams only.

File: src/main/java/com/radixdlt/crypto/ECKeyPair.java (hunk 3 above)
Comment (siy) [reply src/main/java/com/radixdlt/crypto/ECKeyPair.java]: Done

**Pull request merged**